plugins/woocommerce-blocks/assets/js/blocks/product-collection/deprecated.tsx

```diff
@@ -0,0 +1,120 @@
+/**
+ * External dependencies
+ */
+import { BlockInstance, createBlock } from '@wordpress/blocks';
+import { useInnerBlocksProps, useBlockProps } from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import { LayoutOptions, ProductCollectionLayout } from './types';
+
+interface DisplayLayoutAttribute {
+	type: 'list' | 'flex';
+	columns: number;
+	shrinkColumns: boolean;
+}
+
+const migrateDisplayLayout = (
+	attributes: { displayLayout: DisplayLayoutAttribute },
+	innerBlocks: BlockInstance[]
+) => {
+	const { displayLayout = null, ...newAttributes } = attributes;
+	if ( ! displayLayout ) {
+		return [ attributes, innerBlocks ];
+	}
+
+	const { type, columns, shrinkColumns } = displayLayout;
+
+	// Convert custom displayLatout to templateLayout which will then
+	// be applied automatically to the template block.
+	let templateLayout: ProductCollectionLayout;
+	if ( type === 'flex' ) {
+		// Convert responsive columns to size-based automatic columns.
+		if ( shrinkColumns ) {
+			// We can replicate the previous responsive behavior with percentage
+			// based minimum width. One consideration, however, is that the
+			// block gap is not accounted for. We can approximate it by
+			// not using 90% as the width for our column area instead.
+			const columnWidth = Math.floor( 90 / columns );
+			templateLayout = {
+				type: LayoutOptions.GRID,
+				minimumColumnWidth: columnWidth + '%',
+			};
```

Comment:
I went with 90% because it felt like a good approximation. At your prompting, however, I took a deeper look. We dealt with this in the old styles using this CSS:

```scss
$gap-count: calc(#{ $i } - 1);
$total-gap-width: calc(#{ $gap-count } * #{ $grid-gap });
$max-product-width: calc((100% - #{ $total-gap-width }) / #{ $i });

&.columns-#{ $i } {
  grid-template-columns: repeat(auto-fill, minmax(max(#{ $min-product-width }, #{ $max-product-width }), 1fr));
}
```

As we can see, it includes a calculation for the gap width and uses a percentage.

**A caveat, however, is that we're now using a dynamic `gap` value through `supports.spacing.blockGap`. Even _if_ we used this CSS class we wouldn't be able to replicate the previous responsive behavior!**

Using the above calculation and the default `blockGap` value I will take another pass at having the percentage be identical.

Reply:
Thanks again @kmanijak! I went back and adjusted the math. It should now work in every case, please test it out!

Assuming it works, I'll make some changes to the editor controls so that they look correct. As an aside, if we support block spacing, we _can't_ support the kind of responsive layout we had before without writing our own block spacing and abandoning this migration altogether.

This was fine before because the block gap was hardcoded. With it being configurable like this, our CSS classes can't know the size and therefore can't do the necessary math.